### ISSUE TITLE
Sort Performance Degradating when nthreads bigger than machine cores

### ISSUE THREAD

**Issue by chi2liu** [h2oai/datatable#2249]
I wonder why sort operation is sensitive on nthreads.
Test on my machine, when nthreads is just one more than the cores.
it will occur Significant performance degradation

29 threads vs 20 threads 【machine is 28 core】

![image](https://user-images.githubusercontent.com/19861628/71069877-93757600-21b4-11ea-8342-bf095957e7b0.png)

# machine：

cpu：Intel(R) Xeon(R) Gold 5117 CPU @ 2.00GHz
cores：28
processor	: 27
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) Gold 5117 CPU @ 2.00GHz
stepping	: 4
microcode	: 0x2000043
cpu MHz		: 2000.000
cache size	: 19712 KB
physical id	: 1
siblings	: 14
core id		: 14
cpu cores	: 14

MEM: 263412544k

**Comment by st-pasha** [comment]
Using more threads than cores results in oversubscription of CPU resources, which dramatically increases the number of context switches and the resulting execution time. Moreover, the OS won't be able to assign a fixed CPU to each thread any longer, which means that after the context switch a thread is likely to be scheduled on another core, and therefore will have to copy its data over to the other CPU cache, which also contributed to the increase in execution time.

In short, having more worker threads than the number of cores is detrimental for performance, especially for compute- and memory- intensive tasks.

**Comment by chi2liu** [comment]
That makes sense！
Also found another phenomenon: When some system processes on the machine are running all the time, 80% ~ 120% of the CPU will be taken up(our machine is 28 cores).When set to 26 cores, 27 cores and 28 cores, cpu occupancy is generally not full. For example, at 28 cores, it will fluctuate around 2600%.At this time, the performance will be similar to that when the number of cores is exceeded. Can think this is caused by the same cause?

<img width="415" alt="图片" src="https://user-images.githubusercontent.com/19861628/71332882-eb282e80-2572-11ea-9b2a-b3d61e3551fc.png">

In addition, you can see that the performance gradually stabilizes basically between 10 and 20 cores. Can think that the subsequent performance will gradually stabilize with the number of cores, even when switch to a machine with more cores?
By running sort in a single thread, the percentage of the serial part and the parallel part was measured, and the serial part was about 20%. According to Amdahl's law, theoretically the most ideal case is the highest acceleration ratio of 20 cores, 4.17 times. 10-core maximum acceleration ratio, 3.57 times. Referring to the performance of the figure above, you can see that after the 10-core acceleration ratio is stable, the acceleration ratio can reach 3.45. Basically close to the theoretical best performance. However, the number of subsequent cores has increased, and the performance has not significantly improved. Do you think that the current implementation of sort in parallel to multi-core utilization tends to be full at about 10 cores?
